gremlin-scala/src/main/scala/gremlin/scala/ScalaElement.scala

```diff
@@ -14,17 +16,18 @@ trait ScalaElement[ElementType <: Element] {
 
   def keys: Set[String] = element.keys.toSet
 
+  def setProperty(key: String, value: Any): Any
+
   def property[A](key: String): Property[A] = element.property[A](key)
 
-  def properties(wantedKeys: String*): Seq[Property[Any]] = {
-    val requiredKeys = if (wantedKeys.nonEmpty) wantedKeys else keys
-    requiredKeys map property[Any]
-  }.toSeq
+  def properties[A]: Stream[Property[A]]
+
+  def properties[A](wantedKeys: String*): Stream[Property[A]]
 
-  def propertyMap(wantedKeys: String*): Map[String, Any] = {
-    val requiredKeys = if (wantedKeys.nonEmpty) wantedKeys else keys
-    requiredKeys map { key ⇒ (key, getValue(key)) }
-  }.toMap
+  def valueMap[A]: Map[String, A] = valueMap[A](keys.toSeq: _*)
+
+  def valueMap[A](wantedKeys: String*): Map[String, A] =
```

Comment:
i don't think we can return `Map[String, A]` here. what if I'm asking for `valueMap("someInteger", "someString")`? I'll get a Map with Integer and String values, which is a `Map[String, Any]`